src/menus/EditMenus.cpp

```diff
@@ -73,65 +79,173 @@ bool DoPasteText(AudacityProject &project)
    return false;
 }
 
-// Return true if nothing selected, regardless of paste result.
-// If nothing was selected, create and paste into NEW tracks.
+void ForEachCopiedWaveTrack(const TrackList& src,
+                            const TrackList& dst,
+                            const std::function<void(const WaveTrack& waveTrack)>& f)
+{
+   if(dst.Selected().empty())
+   {
+      for(auto waveTrack : src.Any<const WaveTrack>())
+         f(*waveTrack);
+   }
+   else
+   {
+      const auto srcTrackRange = src.Any<const WaveTrack>();
+      const auto dstTrackRange = dst.Any<const WaveTrack>();
+      auto srcTrack = srcTrackRange.begin();
+      auto dstTrack = dstTrackRange.begin();
+      auto lastCopiedTrack = srcTrack;
+      while(dstTrack != dstTrackRange.end() && srcTrack != srcTrackRange.end())
+      {
+         if(!(*dstTrack)->GetSelected())
+         {
+            ++dstTrack;
+            continue;
+         }
+         
+         auto srcChannelCount = TrackList::Channels(*srcTrack).size();
+         auto dstChannelCount = TrackList::Channels(*dstTrack).size();
+                  
+         while(srcChannelCount > 0 && dstChannelCount > 0)
+         {
+            f(**srcTrack);
+            
+            lastCopiedTrack = srcTrack;
+            ++srcTrack;
+            ++dstTrack;
+            --srcChannelCount;
+            --dstChannelCount;
+         }
+         
+         while(dstChannelCount > 0)
+         {
+            f(**lastCopiedTrack);
+            ++dstTrack;
+            --dstChannelCount;
+         }
+      }
+      while(dstTrack != dstTrackRange.end())
+      {
+         if((*dstTrack)->GetSelected())
+            f(**lastCopiedTrack);
+         ++dstTrack;
+      }
+   }
+}
+
+wxULongLong EstimateCopyBytesCount(const TrackList& src, const TrackList& dst)
+{
+   wxULongLong result{};
+   ForEachCopiedWaveTrack(src, dst, [&](const WaveTrack& waveTrack) {
```

Comment:
The algorithm in `ForEachCopiedWaveTrack` might visit one source track multiple times and count its space multiply in the estimate.

This isn't needed for copying and pasting within a project because of the sharing of data.

Rather than all this work, it suggests to me that cross-project copying and pasting needs a smarter algorithm to replicate the graph structure and pattern of sharing.

This could prevent explosion of file sizes in the first place without so much need to warn about it.

Reply:
>The algorithm in ForEachCopiedWaveTrack might visit one source track multiple times and count its space multiply in the estimate.

Also it's possible that different source tracks can reference same blocks